spices/0002-uri-rewriting-for-mirroring.adoc

```diff
@@ -0,0 +1,319 @@
+Note that, because upstream dependencies specify the SHA256 hash, altering packages on a mirror requires _all upstream dependants to be updated_.
+This SPICE does _not_ provide further facilities for altering packages when mirroring.
+
+== Proposed Solution
+
+Similar to `git-config`'s `insteadOf`, the proposed solution is to rewrite URIs when used in `import`, `import*`, `amends`, `extends`, `read`, and `read*` clauses/expressions.
+From `git-config`'s `man` page, on `url.<base>.insteadOf`:
+
+> Any URL that starts with this value will be rewritten to start, instead, with `<base>`.
+> In cases where some site serves a large number of repositories, and serves them with multiple access methods, and some users need to use different access methods, this feature allows people to specify any of the equivalent URLs and have Git automatically rewrite the URL to the best alternative for the particular user, even for a never-before-seen repository on the site.
+> When more than one insteadOf strings match a given URL, the longest match is used.
+>
+> Note that any protocol restrictions will be applied to the rewritten URL.
+> If the rewrite changes the URL to use a custom protocol or remote helper, you may need to adjust the `protocol.*.allow` config to permit the request.
+
+The solution proposed in this SPICE mirrors this design, but extends it to allow regular expressions for the match.
+Only after rewriting are URIs tested for compliance with Allowed Modules and Allowed Resources settings.
+
+Rewrites can be provided:
+
+1. Via the API;
+2. Via `pkl:Project`; and
+3. Via arguments on the CLI.
+
+== Detailed design
+
+=== API
+
+The API changes described here refer to the Java API, but this SPICE implies analogous changes to the Swift and Go APIs.
+
+The `EvaluatorBuilder` gets the following new methods:
+
+[source,java]
+----
+public final class EvaluatorBuilder {
+    // ...
+
+    /**
+     * Adds the given URI rewrite rule.
+     *
+     * @param replace   The literal string to match.
+     * @param with      The literal string to substitute for the first match of [replace].
+     */
+    public EvaluatorBuilder addRewriteRule(String replace, String with) {
+        // ...
+    }
+
+    /**
+     * Adds the given URI rewrite rule.
+     *
+     * @param replace   The regular expression to match.
+     * @param with      The string to substitute for the first match of [replace], which may contain
+     *                  references to capture groups (`$0`, `$1`, etc).
+     */
+    public EvaluatorBuilder addRewriteRule(Pattern replace, String with) {
+        // ...
+    }
+}
+----
+
+=== Standard library module `pkl:Project`
+
+Rewrite rules can be defined project-wide by declaring them in the `PklProject` file.
+
+[source,pkl]
+----
+@Since { version = "0.26.0" }
+class UriRewrite {
+  /// What to match in the given URI.
+  replace: String|Regex
```

Comment:
I think using `pkl.Regex` as a configuration value is an anti-pattern, because it is specifically a Java 8 compatible regex, and not portable across languages.

Following that, I think this should be a `String` instead:

```
/// A regular expression pattern describing what to match in a given URI.
///
/// Regular expressions follow Java rules.
@SourceCode { language = "RegExp" }
replace: String(isRegex)
```

And `with` unconditionally accepts placeholders for capture groups.

Reply:
I was hoping for an easy way to allow non-regex pattern specification, to not have to always have to handle escaping (in the URI case, especially of `.`) and I wasn't considering the case where Pkl's own configuration would be object-mapped to a host language.

Mind you, `Regex` uses `Pattern.compile(pattern, Pattern.UNICODE_CHARACTER_CLASS | Pattern.UNICODE_CASE)`, where `isRegex` uses `Pattern.compile(self, Pattern.UNICODE_CASE)`, so they're the same Java language level, regardless, and the former is slightly more expressive.

Reply:
Yeah, since we are dealing with Java regexes here, it's quite nice that we can use the `isRegex` helper.

The pattern that I was saying we should follow is to use strings to represent regex values, which is a universal way to describe them (`String` is a suitable type no matter the regex flavor).

IMO: `Regex` should only be used for in-language string matching.

Reply:
Users need to understand the exact regex semantics. “follow Java rules” is too vague, and `isRegex` only covers regex syntax. Because the regex will be evaluated by Pkl (even if it happens in Java code), `replace: Regex` seems logical to me. I don’t see why portability matters here.

Reply:
> Users need to understand the exact regex semantics. “follow Java rules” is too vague (and why refer Pkl users to Java

I disagree with this, actually. When writing regex, it's much easier to understand "this is a Java regex" than needing to look through the full set of available. It's not any different than the many tools that refer their users to Perl (see [PCRE](https://en.wikipedia.org/wiki/Perl_Compatible_Regular_Expressions)).

As a matter of principle, `Regex` shouldn't be used to express configuration data in Pkl; it's a data type that's meant for in-language operations. The same can be said of `IntSeq`, `Pair`, `List`, `Map`, etc.